### ISSUE TITLE
Functional Approximation is poor

### ISSUE THREAD

**Issue by 1ssb** [KindXiaoming/pykan#156]
Congratulations on this great piece of work. I have tried to do simple rests like functional approximations and it turns out for a variety of models KANs are extremely poor performing wrt standard MLPs. 

It could be that my model is much simpler and therefore not capable enough, but it would be nice to see these standard comparisons because that would inform how well they can ve adapted broadly.

**Comment by 1ssb** [comment]
Here is a comparison between the models (Same layer configurations -- for both I have used the same positional encodings):

![W B Chart 11_5_2024, 1_18_53 AM](https://github.com/KindXiaoming/pykan/assets/40661648/d6418ca3-fb95-4c9e-8c10-d3d1fb6b48cd)

![W B Chart 11_5_2024, 1_18_33 AM](https://github.com/KindXiaoming/pykan/assets/40661648/f70e27fd-a356-4ce8-bce5-b2dc0cdaadb1)

Both of them are trying to approximate the inverse of the Gaussian model.

**Comment by 1ssb** [comment]
Yes. I did, does not help.

Also since we are trying to move towards interpretability, lets rake some additional onus on ourselves and explain why that would make any difference anyway. Smaller grids would mean more control points right, but would that matter for a smooth function like an inverted gaussian?

**Comment by pop756** [comment]
![image](https://github.com/KindXiaoming/pykan/assets/108378864/2123377f-7dc5-43e0-b3a8-e83be145fe58)
![image](https://github.com/KindXiaoming/pykan/assets/108378864/509c622b-b6ff-4664-831b-98ec38162b83)
If we infer a function such as e^(x^2+1), we divide the sequence into two parts and construct a KAN with two weight layers, x^2 + 1 and exp. If we construct a KAN with a layer such as [1,2,1], we can theoretically predict the above function correctly. However, if we use a different layer, for example, 1,3,1 or 1,3,2,1, the KAN will not make a prediction as shown below, but an additional function will be added when the input value is entered, causing an overfitting phenomenon, and I thought this phenomenon could occur. 

When I configured the KAN layer as [1,5,2,1],grid=2, the result was lower than 0.00002 when training.

The MLP was configured as [1,64,64,1].

**Comment by pop756** [comment]
![image](https://github.com/KindXiaoming/pykan/assets/108378864/091a057f-3492-441d-b4aa-6c4be55e1d78)
![image](https://github.com/KindXiaoming/pykan/assets/108378864/f9852fbc-65a6-491e-8c1c-26dba85d7ab2)



If we infer a function such as e^(x^2+1), we divide the sequence into two parts and construct a KAN with two weight layers, x^2 + 1 and exp. If we construct a KAN with a layer such as [1,2,1], we can theoretically predict the above function correctly. However, if we use a different layer, for example, 1,3,1 or 1,3,2,1, the KAN will not make a prediction as shown below, but an additional function will be added when the input value is entered, causing an overfitting phenomenon, and I thought this phenomenon could occur. 

When I configured the KAN layer as [1,5,2,1],grid=2, the result was lower than 0.00002 when training.

The MLP was configured as [1,64,64,1].

Translated with www.DeepL.com/Translator (free version)

**Comment by pop756** [comment]
![image](https://github.com/KindXiaoming/pykan/assets/108378864/22a4b7f3-ec75-4cf7-85d3-8f0f22249c15)
![image](https://github.com/KindXiaoming/pykan/assets/108378864/02250559-a4ce-443e-840e-d14cea9c0ee4)


If we infer a function such as e^(x^2+1), we divide the sequence into two parts and construct a KAN with two weight layers, x^2 + 1 and exp. If we construct a KAN with a layer such as [1,2,1], we can theoretically predict the above function correctly. However, if we use a different layer, for example, 1,3,1 or 1,3,2,1, the KAN will not make a prediction as shown below, but an additional function will be added when the input value is entered, causing an overfitting phenomenon, and I thought this phenomenon could occur. 

When I configured the KAN layer as [1,5,2,1],grid=2, the result was lower than 0.00002 when training.

The MLP was configured as [1,64,64,1].

**Comment by 1ssb** [comment]
Sorry I should have been clearer. An inverted Gaussian as in f(x) = N(u, sigma^2) then y =f^-1 is what one has access to not a a parameterised weight model as such just saples from the inverse which depending on mapping is an R->R^n function.

**Comment by pop756** [comment]
I'm sorry, I don't have the math knowledge to answer that question, but when I trained the function I presented above, I left the last 20% x (i.e. x corresponding to 1.8-3) as the validation set and trained the remaining 80% x data, the results were as follows when grid =7 and when grid = 3. I didn't apply L1 regulation here. I thought it might help you with the attached question.

Grid = 7
![image](https://github.com/KindXiaoming/pykan/assets/108378864/007eaec1-91bd-4955-84bd-cf1fbf6e4130)
Grid = 3
![image](https://github.com/KindXiaoming/pykan/assets/108378864/330f87ae-eb7b-4e53-99b3-2a2b4cef0c51)

**Comment by 1ssb** [comment]
I am fundamentally unsure of what you mean by divide it into two parts or sequences, can you kindly explain?

**Comment by pop756** [comment]
![image](https://github.com/KindXiaoming/pykan/assets/108378864/c7b134a0-0389-4888-b794-274207f0369b)
This is the result of using this part as training data and
![image](https://github.com/KindXiaoming/pykan/assets/108378864/9d29c848-d83e-4f0a-af31-3254997a3e4c)
the rest of this photo as validation data. And when I trained like that, I changed only the Grid variable and plotted the learning process, I got the following result.

**Comment by 1ssb** [comment]
The errors I got were quite low but they got saturated as in 0.003 but it didnt go below that while MLPs went to the order of 1e-6.
